### PR TITLE
Add logging

### PR DESCRIPTION
**Pull request by andrew-lewin**
This adds logging so we can actually see what the async publisher is up to

@film42 @mmmries @liveh2o

### CODE REVIEW

File: lib/active_publisher/logging.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    def self.initialize_logger(log_target = STDOUT)
+      oldlogger = defined?(@logger) ? @logger : nil
+      @logger = Logger.new(log_target)
+      @logger.level = Logger::INFO
+      oldlogger.close if oldlogger && !$TESTING # don't want to close testing's STDOUT logging
+      @logger
+    end
+
+    def self.logger
+      defined?(@logger) ? @logger : initialize_logger
```
Comment (liveh2o) [inline lib/active_publisher/logging.rb:16]: I'm not exactly sure about this logger setup process.

File: lib/active_publisher/logging.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    end
+
+    def self.logger=(log)
+      @logger = (log ? log : Logger.new("/dev/null"))
```
Comment (liveh2o) [inline lib/active_publisher/logging.rb:20]: I'm curious as to why we would have a setter and getter that both initialize loggers?

Comment (andrew-lewin) [inline lib/active_publisher/logging.rb:20]: I copied this file over from action_subscriber so I don't know a lot of the whys here. I just took a look at action_subscriber and it doesn't look like we are using this setter over there at all. @mmmries, do you know what we'd need to have this setter for or is it safe to take it out?

Comment (liveh2o) [inline lib/active_publisher/logging.rb:20]: Ah, so we have @film42 to blame? :wink:

Comment (mmmries) [inline lib/active_publisher/logging.rb:20]: I know a lot of the considerations around this are for allowing us to set `logger = nil` and have it still work. We actually lifted this code from the `sidekiq` project so I don't know if we have a lot of understanding around all of the nuances ourselves. We were trying to leverage the hard-won experience of the `sidekiq` project without having to learn it all ourselves.

Comment (liveh2o) [inline lib/active_publisher/logging.rb:20]: That's good to know. I seemed like a foreign pattern, which is why I called it out. Perhaps a comment indicating we lifted it from Sidekiq would help people like me from questioning it in the future :sunglasses:

Comment (mmmries) [inline lib/active_publisher/logging.rb:20]: There is a comment at the top of the file `# Taken from https://github.com/mperham/sidekiq/blob/7f882787e53d234042ff18099241403300a47585/lib/sidekiq/logging.rb`

But I think this does bring up a good question. Do we care about allowing `logger` to be nil? If we really wanted to support this use case it seems like we could just catch the `nil` value in the setter and create a `FakeLogger` object that responds to `debug`, `error`, etc but does not actually log anything

Comment (mmmries) [inline lib/active_publisher/logging.rb:20]: Whether we want to support setting `logger = nil` is probably a question for another day.

**Comment by mmmries** [comment]
:shipit: this is definitely a good step forward, we can debate about the best way to handle some of the edge cases in the future.

File: lib/active_publisher/logging.rb (hunk 2 above)
Comment (liveh2o) [inline lib/active_publisher/logging.rb:20]: :+1:

**Pull request merged**